### PR TITLE
Delay a bug when failed to normalize trait ref during specialization

### LINKED ISSUE #43037: ICE in trait normalization when specialization enabled
The following code ([playground](https://is.gd/V3JsSV))

```rust
#![crate_type = "lib"]
#![feature(specialization)] // comment this line to get the expected behavior
trait X {}
trait Y: X {}
trait Z { type Assoc: Y; }
struct A<T>(T);

impl<T> Y for T where T: X {}
impl<T: X> Z for A<T> { type Assoc = T; }

// this impl is invalid, but causes an ICE anyway
impl<T> From<<A<T> as Z>::Assoc> for T {}
```

produces

```
error: internal compiler error: /checkout/src/librustc/traits/specialize/mod.rs:201: failed to fully normalize <T as std::convert::From<<A<T> as Z>::Assoc>>: [FulfillmentError(Obligation(predicate=Binder(TraitPredicate(<T as X>)),depth=1),Unimplemented)]
```

### LINKED ISSUE #45814: specialization involving associated types of unfullfilled traits can cause a panic
At least to my understanding, this bug is caused by using the specialization feature with a trait while using an associated type of a possibly unfullfilled trait in one of the `impl`'s.

I tried this code:

```
#![feature(specialization)]

pub trait Trait<T> { }

default impl<T, U> Trait<T> for U { }

impl<T> Trait<<T as Iterator>::Item> for T { }

fn main() { }
```

I expected to see this happen: An error because T is not necessarily an `Iterator`

Instead, this happened: `rustc` just panics

## Meta

`rustc --version --verbose`:

```
rustc 1.23.0-nightly (d762b1d6c 2017-11-04)
binary: rustc
commit-hash: d762b1d6c67db12e117186d94d70e46cddb22965
commit-date: 2017-11-04
host: x86_64-unknown-linux-gnu
release: 1.23.0-nightly
LLVM version: 4.0
```

Backtrace:

Quite large, I put it in a file for that reason (56 steps): 
[backtrace.txt](https://github.com/rust-lang/rust/files/1447751/backtrace.txt)

### PR DESCRIPTION
**Pull request by compiler-errors**
The error messages still kinda suck here but they don't ICE anymore...

Fixes #45814
Fixes #43037

r? types

### CODE REVIEW

File: compiler/rustc_trait_selection/src/traits/specialize/mod.rs [hunk 1]
```diff
@@ -159,8 +159,12 @@ pub(super) fn specializes(tcx: TyCtxt<'_>, (impl1_def_id, impl2_def_id): (DefId,
             impl1_trait_ref,
         ) {
             Ok(impl1_trait_ref) => impl1_trait_ref,
-            Err(err) => {
```
Comment (b-naber) [inline compiler/rustc_trait_selection/src/traits/specialize/mod.rs:162]: Can we remove this comment above?

```
        // Normalize the trait reference. The WF rules ought to ensure
        // that this always succeeds.
```

**Comment by spastorino** [comment]
@compiler-errors let me know when you address @b-naber comments and I can merge this.

**Comment by compiler-errors** [comment]
Removed the comment.

**Pull request merged**